File: 2023/day01/main.hs
import Data.Char (isDigit)
import Data.Function (on)
import Data.List (findIndex, inits, isPrefixOf, isSuffixOf, maximumBy, minimumBy, tails)
import Data.Ord (comparing)
import Data.Maybe (isJust, fromJust)

part1 :: IO ()
part1 = do
  contents <- readFile "inputs/main.input"
  let linesOfFile = lines contents
  let digits = map mapDigitToPart1 linesOfFile
  print (sum digits)

mapDigitToPart1 :: String -> Int
mapDigitToPart1 xs = read (head [[x] | x <- xs, isDigit x]) * 10 + read (last [[x] | x <- xs, isDigit x]) :: Int

part2 :: IO ()
part2 = do
  contents <- readFile "inputs/main.input"
  let linesOfFile = lines contents
  print (sum (map mapDigitToPart2 linesOfFile))

numbersList = [("one", 1), ("1", 1), ("two", 2), ("2", 2), ("three", 3), ("3", 3), ("four", 4), ("4", 4), ("five", 5), ("5", 5), ("six", 6), ("6", 6), ("seven", 7), ("7", 7), ("eight", 8), ("8", 8), ("nine", 9), ("9", 9)]

mapDigitToPart2 :: String -> Int
mapDigitToPart2 xs = snd (findFirstDigit xs) * 10 + snd (findLastDigit xs)

findFirstDigit :: String -> (Int, Int)
findFirstDigit str = minimumBy (comparing fst) [(fromJust (findSubstring ptrn str), value) | (ptrn, value) <- numbersList, isJust (findSubstring ptrn str)]

findSubstring :: (Eq a) => [a] -> [a] -> Maybe Int
findSubstring pat str = findIndex (isPrefixOf pat) (tails str)

findLastDigit :: String -> (Int, Int)
findLastDigit str = minimumBy (comparing fst) [(fromJust (findRSubstring pat str), value) | (pat, value) <- numbersList, isJust (findRSubstring pat str)]

findRSubstring :: (Eq a) => [a] -> [a] -> Maybe Int
findRSubstring ptrn str = findIndex (isSuffixOf ptrn) (reverse (inits str))

File: 2023/day01/src/main.rs
use std::error::Error;

use advent_of_code_lib::{self, Solver};
use dotenv::dotenv;

fn main() -> Result<(), Box<dyn Error>> {
    dotenv().ok();
    advent_of_code_lib::run_and_print(Day, "2023", "1")
}

struct Day;

const NUMBERS: &[(&str, u32)] = &[
    ("0", 0),
    ("zero", 0),
    ("1", 1),
    ("one", 1),
    ("2", 2),
    ("two", 2),
    ("3", 3),
    ("three", 3),
    ("4", 4),
    ("four", 4),
    ("5", 5),
    ("five", 5),
    ("6", 6),
    ("six", 6),
    ("7", 7),
    ("seven", 7),
    ("8", 8),
    ("eight", 8),
    ("9", 9),
    ("nine", 9),
];

impl Solver for Day {
    fn part1(&self, input: &str) -> String {
        input
            .lines()
            .map(|x| {
                let min = x
                    .chars()
                    .filter(|char| char.is_digit(10))
                    .next()
                    .unwrap()
                    .to_digit(10)
                    .unwrap();

                let max = x
                    .chars()
                    .filter(|char| char.is_digit(10))
                    .last()
                    .unwrap()
                    .to_digit(10)
                    .unwrap();
                min * 10 + max
            })
            .sum::<u32>()
            .to_string()
    }

    fn part2(&self, input: &str) -> String {
        input
            .lines()
            .map(|x| {
                let min = NUMBERS
                    .iter()
                    .filter_map(|(st, value)| Some((x.find(st)?, value)))
                    .min_by_key(|x| x.0)
                    .unwrap();
                let max = NUMBERS
                    .iter()
                    .filter_map(|(st, value)| Some((x.rfind(st)?, value)))
                    .max_by_key(|x| x.0)
                    .unwrap();
                min.1 * 10 + max.1
            })
            .sum::<u32>()
            .to_string()
    }
}
